Source/carthage/Outdated.swift

```diff
@@ -87,8 +89,18 @@ public struct OutdatedCommand: CommandProtocol {
 	public let function = "Check for compatible updates to the project's dependencies"
 
 	public func run(_ options: Options) -> Result<(), CarthageError> {
+        let shouldIgnoreFetching = shouldIgnoreFetchingDependencies(fetchInterval: TimeInterval(options.fetchInterval))
 		return options.loadProject()
-			.flatMap(.merge) { $0.outdatedDependencies(options.isVerbose) }
+            .combineLatest(with: shouldIgnoreFetching)
+            .flatMap(.merge) { project, shouldIgnoreFetching -> SignalProducer<[Project.OutdatedDependency], CarthageError> in
+                if shouldIgnoreFetching {
+                    // TODO: Non-fetching changes required.
+                    carthage.println("The dependencies will not be fetched due to non-expired fetch interval.")
+                    return project.outdatedDependencies(options.isVerbose)
```

Comment:
I noticed that we do download the remote dependencies when we initialise `ResolverProtocol` type object:
https://github.com/Carthage/Carthage/blob/1c3e23a3dffc7f19ccd6dd8073454bb39fdfafa5/Source/CarthageKit/Project.swift#L505-L509

Wondering if modifying the resolver initialisation is a way to go, i.e. passing `[:]` to resolver's versions 🤔 

Could you guys please give me a hint what would be the best way to ignore fetching the remote dependencies? 🙏

Reply:
I think you can just set `FetchCache.fetchCacheInterval` to the value that's passed in. That should be _all_ you need to do?

Reply:
👋 @mdiep Thanks for the comment and your help 🙇 Just to double check: `FetchCache.fetchCacheInterval` is just a static property without the persistent storage behind it, so we need the `UserDefaults` logic I introduced in this PR anyway, right? 

With the suggested changes we are going to have:

```Swift
...
if shouldIgnoreFetching {
    carthage.println("The dependencies will not be fetched due to non-expired fetch interval.")
    FetchCache.fetchCacheInterval = TimeInterval(options.fetchInterval)
    return project.outdatedDependencies(options.isVerbose)
} 
....
```

Could you please just confirm we are on the same page? 🙏

Reply:
> so we need the UserDefaults logic I introduced in this PR anyway, right?

Why?

Reply:
Please correct me if I am wrong @mdiep , but I though `carthage` (CLI application) process is killed as soon as the command execution ends. It means `carthage` does not hold in memory state (the static properties) between the command executions. That's why I introduced `UserDefaults` - to keep the state of the application between the command executions.

Reply:
Oh, right. Sorry, I wasn't thinking.

I'm not sure how we'd want to store that. [This](https://stackoverflow.com/a/9229377) might help?

Reply:
Thanks for the suggestion 🙇I tested the command `stat -f '%m' .git/FETCH_HEAD`, unfortunately **`FETCH_HEAD` object is not updated when `carthage` fetches just the dependencies**:

<img width="678" alt="screenshot 2019-03-04 at 22 13 08" src="https://user-images.githubusercontent.com/10795657/53763949-4d248400-3ecc-11e9-968a-336d48b92b84.png">

The `FETCH_HEAD` gets updated when we do fetch the current project:

![image](https://user-images.githubusercontent.com/10795657/53764041-80671300-3ecc-11e9-8e29-b6bdf08c4729.png)

It seems this idea won't work for us.

Reply:
I was testing `FetchCache` structure by running the same `carthage` command `carthage outdated --xcode-warnings` a few times in a row and `lastFetchTimes` dictionary was empty each time. The abstraction seems to be broken. Maybe there's some context I am missing here? 🤔 

I'd suggest to fix it by introducing the persistent storage, i.e. `UserDefaults` (not sure how that might affect the performance) or finding another way how to keep a state between the command executions. By making this change indeed we could just set the `fetchCacheInterval` to the value passed by `--fetch-interval` parameter.

Reply:
The `FetchCache` is only used for in-process caching right now; it's not persistent.

I don't love using `UserDefaults`. I think we should either infer it from the dependencies directory or store it somewhere on disk there.

Reply:
I see 👍 I will try to find a way to infer the last modification date (last fetch) from the dependencies directory. However I'd like to know why you prefer storing `fetch-interval` value somewhere on disk (in a file/plist probably) instead of using `UserDefaults`? 🤔 

`UserDefaults` are pretty fast and easy to do - the simplest key-value storage available.

Reply:
It's easier to manual trigger. It seems more straightforward if you delete the cache entirely. Even though it's a little harder to implement, I think it's the simpler solution.

Reply:
> It seems more straightforward if you delete the cache entirely.

Could you please explain me a bit more what do you mean by ☝️ 🙏  

I am a bit confused now and dunno in what direction to push this PR.

**First option**

Check if we are able to detect when a last fetch happened based on the `Dependencies` catalog. In that way we wouldn't need to store `fetch-interval` value anywhere, but read it on the runtime. 

**Second option** 

Please correct me if I am wrong - you'd like to keep `fetch-interval` value in the file on the disk and if this param wasn't provided, we would just delete the whole file? 🤔 

🙇 🙇

Reply:
I prefer option one.

But failing that, I think we should keep the last fetched time in a file on disk in the dependencies directory. Users sometimes delete the entire cache to get back to a clean state. I'd prefer to keep all the cache info in that directory instead of putting some of it in user defaults.

Reply:
👋 @mdiep 

I was investigating option one and unfortunately couldn't find a reasonable solution. 

I was playing with with the option number two - file solution. I'd suggest to modify `FetchCache` component and convert it from in-process caching to persistent caching. We could store last fetch timestamp per dependency (`../Library/Caches/org.carthage.CarthageKit/dependencies`): 

![image](https://user-images.githubusercontent.com/10795657/55293472-356eeb80-53f7-11e9-85e9-ca1f38c82115.png)

Please tell me what you think about this solution 🙇

Reply:
What if _we_ set the last modified date of the directory for the dependency?

Reply:
Good idea, thanks! 👍 

Tested on just three dependencies, but already noticed a great improvement (a few seconds):

<img width="1136" alt="Screenshot 2019-04-01 at 08 03 12" src="https://user-images.githubusercontent.com/10795657/55307572-35eb9e80-5458-11e9-9565-a6338713630c.png">

<img width="1136" alt="Screenshot 2019-04-01 at 08 03 27" src="https://user-images.githubusercontent.com/10795657/55307579-3b48e900-5458-11e9-80bc-684ff5768e64.png">

I implemented the changes and kindly ask you for re-review @mdiep  🙇 🙇

Reply:
👋 Hey @mdiep 🙂 Just wanted to kindly ask if you did get a chance to have a look at the newest changes? 🙇 🤔